File: frontend/src/utils/galleryDownload.js
/* ⬇ Taking pictures OFF the board — the decidable half.
 *
 * The ◉ Canvas is the only screen that knows an image's whole ancestry
 * (dataset → run → checkpoint → seed) and, until now, the only way to keep one
 * was a right-click "Save image as…" that threw all of it away. Two things ship:
 * a single image, from the pinned node and from the viewer; and a whole
 * gallery — a run's, or one checkpoint's — as a ZIP.
 *
 * THE NAME IS THE FEATURE and it is NOT computed here. There is no settings
 * sidecar (deliberately), so the file name is the sole carrier of the lineage,
 * and a scheme implemented twice is a scheme that will disagree with itself:
 * it lives once, in `services/gallery_download.py`, where pytest can reach
 * every hostile dataset name. The browser reads the finished name off
 * Content-Disposition or simply lets the download go.
 *
 * What lives HERE is the other half, and it is the half a JSX file would hide:
 * WHICH images a click is about, and whether the screen states the cut BEFORE
 * the archive lands. A ZIP that quietly holds 500 of 812 is exactly the kind of
 * half-truth the rest of this app refuses, so the count is on the button.
 *
 * JSX-free on purpose — `node --test` cannot parse JSX.
 */

/** Hard ceiling on one archive, mirrored from services/gallery_download.py
 *  (ZIP_IMAGE_CAP). Duplicated as a NUMBER only so the button can say the limit
 *  before the round-trip; the backend remains the one that enforces it. */
const ZIP_IMAGE_CAP = 500;

export const ZIP_SCOPE_ALL = 'all';
export const ZIP_SCOPE_SELECTION = 'selection';

function scopePath(target) {
  if (!target) return null;
  // The app-wide 🖼 Gallery: no record to scope by. Its ZIP routes are
  // selection-ONLY (the backend refuses a missing `ids`), so callers there
  // always pass one.
  if (target.kind === 'app') return '/api/gallery/images';
  if (target.recordId == null) return null;
  const rid = target.recordId;
  // Same rule as galleryScope(): no step means the whole run.
  return (target.kind === 'run' || target.step == null)
    ? `/api/train/run/${rid}/images`
    : `/api/train/checkpoint/${rid}/${target.step}/images`;
}

/** `?ids=` for an explicit selection. An EMPTY array keeps the parameter: an
 *  empty selection must stay an empty selection all the way to the backend,
 *  which refuses it — degrading it into "no parameter" would hand the whole
 *  gallery to a click that meant "these ones". */
function idsQuery(ids) {
  if (ids == null) return '';
  return `?ids=${ids.join(',')}`;
}

/** The ZIP itself. `ids` omitted → the whole scope. */
export function galleryZipUrl(target, ids = null) {
  const base = scopePath(target);
  return base ? `${base}/zip${idsQuery(ids)}` : null;
}

/** The preflight: counts, missing files and the cap, with no byte moved. */
export function galleryZipPlanUrl(target, ids = null) {
  const base = scopePath(target);
  return base ? `${base}/zip/plan${idsQuery(ids)}` : null;
}

/** One image, under its lineage name. */
export function imageDownloadUrl(imageId) {
  return imageId == null ? null : `/api/train/image/${imageId}/download`;
}

/**
 * The name the saved file lands under, read off `Content-Disposition`.
 *
 * NOT rebuilt in the browser: the lineage scheme lives once, in
 * services/gallery_download.py, where pytest can throw hostile dataset names at
 * it. Flask writes the RFC 5987 form beside an ASCII fallback whenever the name
 * needed escaping, and the escaped one has to win — matching the first
 * `filename` in the string would silently pick the lossy one.
 */
export function nameFromDisposition(header, fallback = 'image.png') {
  const value = String(header || '');
  const star = value.match(/filename\*=UTF-8''([^;]+)/i);
  if (star) {
    try { return decodeURIComponent(star[1].trim()); } catch { /* fall through */ }
  }
  const plain = value.match(/filename="?([^";]+)"?/i);
  return plain ? plain[1].trim() : fallback;
}

/**
 * ⬇ The selection as PLAIN FILES — one save per picked image, no archive.
 *
 * The ZIP next to it answers "give me these to move around as one thing"; this
 * answers "put these in my Downloads as themselves", which is what a phone's
 * photo picker, a training tool watching a folder, or someone grabbing three
 * pictures actually wants — unzipping on a phone is a chore that costs more
 * than the download. Both reuse the SAME single-image route, so every file
 * lands under its lineage name from Content-Disposition; nothing here invents
 * a second naming path.
 *
 * SEQUENTIAL on purpose. Firing N programmatic saves in one tick is how
 * browsers decide you are an attack: Chrome swallows everything past the first
 * ten or so unless the user grants "download multiple files", and the grants
 * prompt fires mid-burst so half the burst is already lost. One at a time, each
 * save waiting for its bytes, stays inside every browser's tolerance — the
 * permission prompt still appears once (that is the browser talking, not us)
 * and nothing is lost while it waits.
 *
 * A missing file SKIPS, it does not stop: the gallery legitimately lists rows
 * whose file a resume or a trash sweep took off the disk, and one such row must
 * not cost the user the other eleven. The summary names the count either way.
 *
 * `deps` exists for `node --test`: fetch and the save step are injectable, so
 * the loop's order, its skip-on-404 and its stop flag are pinned without a DOM.
 */
export async function downloadImagesAsFiles(ids, {
  onProgress = () => {},
  shouldStop = () => false,
  fetchImpl = null,
  saveBlob = null,
} = {}) {
  const doFetch = fetchImpl || ((url) => fetch(url, { credentials: 'same-origin' }));
  const doSave = saveBlob || ((blob, name) => {
    const href = URL.createObjectURL(blob);
    const a = document.createElement('a');
    a.href = href;
    a.download = name;
    document.body.appendChild(a);
    a.click();
    a.remove();
    // Next tick: revoking synchronously races Safari's own read of the blob
    // and the save silently produces nothing (same rule as useImageDownload).
    setTimeout(() => URL.revokeObjectURL(href), 0);
  });
  let saved = 0;
  let skipped = 0;
  const total = ids.length;
  for (const id of ids) {
    if (shouldStop()) break;
    onProgress({ done: saved + skipped, total });
    try {
      const res = await doFetch(imageDownloadUrl(id));
      if (!res.ok) { skipped += 1; continue; }
      const name = nameFromDisposition(res.headers.get('Content-Disposition'));
      await doSave(await res.blob(), name);
      saved += 1;
    } catch {
      skipped += 1;
    }
  }
  return { saved, skipped, stopped: shouldStop() ? total - saved - skipped : 0 };
}

/** The sentence the bar shows when the loop ends. Honest about every branch:
 *  a run someone cancelled mid-way says so instead of pretending it finished. */
export function filesDownloadSummary({ saved, skipped, stopped }) {
  const parts = [`${saved} file${saved === 1 ? '' : 's'} saved`];
  if (skipped) parts.push(`${skipped} skipped (file no longer on disk)`);
  if (stopped) parts.push(`${stopped} not started`);
  return parts.join(', ');
}

/**
 * Everything the ⬇ button in the gallery's action bar shows and does.
 *
 * ONE button, two meanings, taken from the mode already on screen: outside
 * Select mode it is the whole gallery, inside it is the picks. Two buttons both
 * saying "download" and differing by a word is how a bar becomes unreadable —
 * and Select mode is right there, already the app's answer to "these ones".
 */
export function zipButtonState({ picking = false, selectedCount = 0, totalCount = 0,
  cap = ZIP_IMAGE_CAP, busy = false } = {}) {
  const total = Math.max(0, Number(totalCount) || 0);
  const picked = Math.max(0, Number(selectedCount) || 0);
  const scope = picking ? ZIP_SCOPE_SELECTION : ZIP_SCOPE_ALL;
  if (total === 0) {
    return { shown: false, scope, disabled: true, capped: false, count: 0,
      label: '', title: '' };
  }
  const wanted = picking ? picked : total;
  const count = Math.min(wanted, cap);
  const capped = !picking && total > cap;
  let title;
  if (picking) {
    title = picked === 0
      ? 'Pick the images you want, then download them as one ZIP'
      : `Download the ${picked} selected image(s) as one ZIP`;
  } else if (capped) {
    // The cut, on the tooltip AND in the count on the face of the button, so it
    // cannot be discovered by counting files inside the archive.
    title = `Download the newest ${cap} of ${total} images as one ZIP — `
      + `one archive holds at most ${cap}. Open a single checkpoint's gallery `
      + 'for the rest.';
  } else {
    title = `Download all ${total} image(s) of this gallery as one ZIP`;
  }
  return {
    shown: true,
    scope,
    count,
    capped,
    disabled: busy || count === 0,
    label: busy ? '⬇ Zipping…' : `⬇ ZIP (${count})`,
    title,
  };
}

/**
 * The sentence the panel shows once the preflight answers — or the refusal.
 *
 * `null` means "nothing worth saying": a clean, complete, untruncated archive
 * needs no commentary. Everything else does, because the ZIP is about to be
 * shorter than the gallery it came from.
 */
export function planNotice(plan) {
  if (!plan) {
    return { blocked: true, kind: 'error',
      text: 'Could not work out what to download — the gallery could not be read.' };
  }
  if (!plan.ok || !plan.included) {
    return { blocked: true, kind: 'error',
      text: plan.note || 'None of these images can be downloaded.' };
  }
  if (plan.missing > 0 || plan.truncated) {
    return { blocked: false, kind: 'warn', text: plan.note || '' };
  }
  return null;
}

File: backend/app/services/gallery_download.py
"""⬇ Getting generated images OFF the board — one at a time, or a gallery as ZIP.

WHY THE FILE NAME IS THE FEATURE
--------------------------------
The ◉ LoRA Canvas is the only screen in the app that knows an image's whole
ancestry: which dataset, which run, which checkpoint, which seed.  Until now
that knowledge could only be *looked at*.  The moment a picture leaves for the
Downloads folder it keeps whatever ComfyUI happened to call it —
``out_00042_.png`` — and a week later nobody can say which checkpoint produced
it.  Two files from two checkpoints of the same run are then indistinguishable,
which defeats the only reason the board exists.

A settings sidecar (a ``.txt``/``.json`` beside each image) was deliberately
ruled out.  So **the file name is the sole carrier of the lineage**, and it has
to earn that:

    ``<dataset>_run<id>_step<NNNNNN>_seed<seed>_<imageId>.<ext>``
    e.g. ``Nova-Style_run42_step002500_seed208607443_1187.png``

  * **zero-padded step** — the whole point.  Sorted as text, ``2500`` comes
    before ``500``; ``002500`` does not.  Six digits covers every realistic
    schedule, and a step above that simply stops being padded rather than being
    truncated into a lie.
  * **``stepunknown`` for a step-less save** — ``u`` sorts after the digits, so
    the group the run gallery puts last lands last here too, for free.
  * **the image id last** — the collision breaker.  Two renders of the same
    checkpoint with the same seed are a normal thing (different sampler, another
    day); the row id is unique by construction, so the name is too.
  * **the seed segment disappears when there is no seed** rather than becoming
    ``seed None``.
  * portable: NTFS's forbidden set ``<>:"/\\|?*`` never survives the slug,
    neither do control bytes, trailing dots/spaces or the DOS device names, and
    the whole thing is bounded well under the 255-byte component limit that
    Windows, ext4 and APFS all share.

WHY NOT DEFLATE
---------------
Every one of these files is a PNG or a JPEG — already compressed.  Deflating
them again buys a percent or two for a lot of CPU on a machine that is usually
also training something, so the archive is STORED.  A gallery ZIP is then
roughly a disk copy.

CAPS ARE ANNOUNCED, NEVER SILENT
--------------------------------
``gallery_download_plan`` is the cheap half — rows and ``os.path.exists`` only,
no bytes moved — so the UI can say *before* the click how many images it is
about to get, how many files have vanished, and whether the cap cut the list.
A ZIP that quietly holds 500 of 812 images is exactly the kind of half-truth
this project removes everywhere else.
"""
from __future__ import annotations

import os
import re
import zipfile

from .dataset_storage import dataset_path

# How many images one archive may hold.  The gallery panel already caps its
# PAYLOAD at 240; this is the harder, byte-moving limit and it sits at the same
# 500 the checkpoint gallery uses for its own list.  Above it the request would
# hold a worker for minutes on a machine that is usually also busy.
ZIP_IMAGE_CAP = 500

# Bounded so the whole name stays far under the 255-byte component limit shared
# by NTFS, ext4 and APFS — a dataset called "…" 400 characters long must not
# produce a file nobody can save.
MAX_SLUG_LEN = 40
MAX_NAME_LEN = 120

# The names Windows still refuses whatever the extension, because they are
# devices.  A generated name always continues with `_run…`, so this can only
# bite through the slug — it is checked on the finished stem anyway.
RESERVED_DEVICE_NAMES = frozenset(
    ['CON', 'PRN', 'AUX', 'NUL']
    + [f'COM{i}' for i in range(1, 10)]
    + [f'LPT{i}' for i in range(1, 10)])

_SLUG_KEEP = re.compile(r'[^A-Za-z0-9]+')
_EXT_OK = re.compile(r'^[a-z0-9]{1,5}$')


def slugify(text, fallback='dataset'):
    """A file-name-safe fragment of `text`, or `fallback` if nothing survives.

    Deliberately ASCII-only: a name that round-trips through a ZIP, a shell, a
    Discord upload and someone else's file manager cannot depend on the
    encoding any of them chose.  A fully non-Latin dataset name therefore keeps
    its identity through the run id rather than through its letters.
    """
    slug = _SLUG_KEEP.sub('-', str(text or '')).strip('-')
    slug = slug[:MAX_SLUG_LEN].strip('-')
    return slug or fallback


def _extension(filename):
    """The stored file's extension, lowercased and vetted. Anything odd becomes
    `png`: the extension decides what opens the file, so it is the one segment
    that must never carry user-influenced text."""
    ext = os.path.splitext(str(filename or ''))[1].lstrip('.').lower()
    return ext if _EXT_OK.match(ext) else 'png'


def _step_segment(step):
    if step is None:
        return 'stepunknown'
    try:
        n = int(step)
    except (TypeError, ValueError):
        return 'stepunknown'
    return f'step{n:06d}' if 0 <= n < 1000000 else f'step{n}'


def image_download_name(row, dataset_name=None):
    """The lineage-carrying name for ONE generated image row.

    `row` is a ``LoraTestImage``.  `dataset_name` is passed in rather than
    fetched so a ZIP of 500 images does not run 500 dataset lookups.
    """
    slug = slugify(dataset_name, f'dataset{row.dataset_id}')
    parts = [slug, f'run{row.record_id}' if row.record_id is not None else 'rununknown',
             _step_segment(row.step)]
    if row.seed is not None:
        parts.append(f'seed{row.seed}')
    parts.append(str(row.id))
    stem = '_'.join(parts)
    ext = _extension(row.filename)
    # Trim from the SLUG, never from the tail: run / step / id are the lineage
    # and dropping them would leave a long pretty name that says nothing.
    over = len(stem) + 1 + len(ext) - MAX_NAME_LEN
    if over > 0:
        slug = slug[:max(1, len(slug) - over)].strip('-') or 'dataset'
        stem = '_'.join([slug] + parts[1:])
    if stem.upper() in RESERVED_DEVICE_NAMES:
        stem = f'_{stem}'
    return f'{stem}.{ext}'


def _rows_for(record_id, step, image_ids=None):
    """The finished, file-backed rows of this scope, newest first.

    Scoped exactly like ``delete_checkpoint_images``: ``step=None`` means the
    whole run, and an id belonging to another run is simply not found.  Reusing
    the delete's shape is on purpose — a download that could reach rows the
    delete cannot would be a second, quieter definition of "this gallery".
    """
    from ..models import LoraTestImage
    q = LoraTestImage.query.filter(LoraTestImage.record_id == record_id,
                                   LoraTestImage.status == 'done',
                                   LoraTestImage.filename.isnot(None))
    if step is not None:
        q = q.filter(LoraTestImage.step == step)
    if image_ids is not None:
        wanted = []
        for i in image_ids:
            try:
                wanted.append(int(i))
            except (TypeError, ValueError):
                continue
        if not wanted:
            return []
        q = q.filter(LoraTestImage.id.in_(wanted))
    return q.order_by(LoraTestImage.id.desc()).all()


def _dataset_names(rows):
    from ..models import FaceDataset
    ids = {r.dataset_id for r in rows if r.dataset_id is not None}
    if not ids:
        return {}
    return {d.id: d.name
            for d in FaceDataset.query.filter(FaceDataset.id.in_(ids)).all()}


def zip_download_name(record_id, step, dataset_name, partial=False):
    """What the archive itself is called — the same grammar as its contents."""
    slug = slugify(dataset_name, f'dataset{record_id}')
    tail = 'selection' if partial else 'gallery'
    scope = '' if step is None else f'_{_step_segment(step)}'
    return f'{slug}_run{record_id}{scope}_{tail}.zip'


def _note(total, included, missing, truncated, cap):
    """The sentence the screen shows. Every cut is named here, once, so the
    button, the toast and the tests cannot drift into three different stories."""
    if total == 0:
        return 'There is nothing to download here yet.'
    if included == 0:
        return (f'None of these {total} images can be downloaded — '
                f'{missing} file(s) are no longer on disk.')
    bits = []
    if truncated:
        bits.append(f'Downloading the newest {included} of {total} '
                    f'(one archive holds at most {cap}).')
    else:
        bits.append(f'Downloading {included} image(s).')
    if missing:
        bits.append(f'{missing} file(s) are no longer on disk and were left out.')
    return ' '.join(bits)


def gallery_download_plan(record_id, step, image_ids=None, cap=ZIP_IMAGE_CAP):
    """What a ZIP of this scope WOULD hold — computed without moving a byte.

    Returns ``{ok, total, included, missing, truncated, cap, note, filename,
    entries}`` where each entry is ``{id, path, name}``.  `ok` is False when
    there is nothing left to put in the archive, which is the case the button
    has to refuse out loud rather than hand over an empty ZIP.
    """
    cap = max(1, int(cap or ZIP_IMAGE_CAP))
    rows = _rows_for(record_id, step, image_ids)
    total = len(rows)
    kept = rows[:cap]                      # newest first — the panel's own rule
    truncated = total > len(kept)
    names = _dataset_names(kept)
    entries, missing, seen = [], 0, set()
    for row in kept:
        path = os.path.join(dataset_path(row.dataset_id), row.filename)
        if not os.path.exists(path):
            missing += 1
            continue
        name = image_download_name(row, names.get(row.dataset_id))
        # Belt and braces: the row id is already in the name, so this can only
        # fire if the scheme ever changes underneath us.
        if name in seen:
            name = f'{row.id}_{name}'
        seen.add(name)
        entries.append({'id': row.id, 'path': path, 'name': name})
    included = len(entries)
    first = kept[0] if kept else None
    ds_name = names.get(first.dataset_id) if first is not None else None
    return {
        'ok': included > 0,
        'total': total,
        'included': included,
        'missing': missing,
        'truncated': truncated,
        'cap': cap,
        'note': _note(total, included, missing, truncated, cap),
        'filename': zip_download_name(record_id, step, ds_name,
                                      partial=image_ids is not None),
        'entries': entries,
    }


def write_gallery_zip(entries, output):
    """Write the planned entries into `output`, sorted by their names.

    STORED, not deflated: these are PNG/JPEG bytes and re-compressing them costs
    CPU for nothing.  Sorted, so the archive opens in training order — which is
    exactly what the zero-padded step in the name buys.
    """
    with zipfile.ZipFile(output, 'w', zipfile.ZIP_STORED) as zf:
        for e in sorted(entries, key=lambda x: x['name']):
            try:
                zf.write(e['path'], arcname=e['name'])
            except OSError:
                # A file that vanished between the plan and the write. Skipping
                # is right — the plan already reported what it found, and one
                # unreadable file must not lose the other 499.
                continue


def single_image_download(image_id):
    """``(path, name)`` for one generated image, or ``(None, reason)``.

    A plain ``<a download>`` on the existing image URL would have been shorter
    and is exactly what this avoids: when the file is gone the browser saves the
    404 page under the .png name and the user finds out by opening it. Resolving
    here means the caller gets a refusal it can say out loud.
    """
    from ..extensions import db
    from ..models import FaceDataset, LoraTestImage
    row = db.session.get(LoraTestImage, image_id)
    if row is None or not row.filename:
        return None, 'That image is no longer in the library.'
    path = os.path.join(dataset_path(row.dataset_id), row.filename)
    if not os.path.exists(path):
        return None, 'That image file is no longer on disk.'
    ds = db.session.get(FaceDataset, row.dataset_id) if row.dataset_id else None
    return path, image_download_name(row, ds.name if ds else None)
